File: backend/app/api/v1/notifications.py
"""Notification REST endpoints (Phase 10)."""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.dependencies.db import get_db
from app.schemas.common import DataResponse, ListResponse, PaginationMeta
from app.schemas.notification import (
    NotificationCreate,
    NotificationRead,
    NotificationUpdate,
    UnreadCountRead,
)
from app.services.notification import NotificationService

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=ListResponse[NotificationRead])
def list_notifications(
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=25, ge=1, le=100),
    unread_only: bool = Query(default=False),
    recipient: str | None = Query(default=None, max_length=320),
    kind: str | None = Query(default=None, max_length=32),
    db: Session = Depends(get_db),
) -> ListResponse[NotificationRead]:
    service = NotificationService(db)
    items, total = service.list(
        page=page,
        limit=limit,
        unread_only=unread_only,
        recipient=recipient,
        kind=kind,
    )
    return ListResponse(
        data=[service.to_read(item) for item in items],
        pagination=PaginationMeta.from_totals(page=page, limit=limit, total=total),
    )


@router.get("/unread-count", response_model=DataResponse[UnreadCountRead])
def unread_count(
    recipient: str | None = Query(default=None, max_length=320),
    db: Session = Depends(get_db),
) -> DataResponse[UnreadCountRead]:
    count = NotificationService(db).unread_count(recipient=recipient)
    return DataResponse(data=UnreadCountRead(count=count))


@router.post("/read-all", response_model=DataResponse[UnreadCountRead])
def mark_all_read(
    recipient: str | None = Query(default=None, max_length=320),
    db: Session = Depends(get_db),
) -> DataResponse[UnreadCountRead]:
    count = NotificationService(db).mark_all_read(recipient=recipient)
    return DataResponse(
        data=UnreadCountRead(count=count),
        message="Marked notifications as read.",
    )


@router.post(
    "",
    response_model=DataResponse[NotificationRead],
    status_code=status.HTTP_201_CREATED,
)
def create_notification(
    payload: NotificationCreate,
    db: Session = Depends(get_db),
) -> DataResponse[NotificationRead]:
    """Create a notification (system/admin path). Assignment alerts are automatic."""
    service = NotificationService(db)
    notification = service.create(payload)
    return DataResponse(data=service.to_read(notification))


@router.get("/{notification_id}", response_model=DataResponse[NotificationRead])
def get_notification(
    notification_id: UUID,
    db: Session = Depends(get_db),
) -> DataResponse[NotificationRead]:
    service = NotificationService(db)
    notification = service.get(notification_id)
    return DataResponse(data=service.to_read(notification))


@router.patch("/{notification_id}", response_model=DataResponse[NotificationRead])
def update_notification(
    notification_id: UUID,
    payload: NotificationUpdate,
    db: Session = Depends(get_db),
) -> DataResponse[NotificationRead]:
    """Mark a notification read or unread."""
    service = NotificationService(db)
    notification = service.mark_read(notification_id, read=payload.read)
    return DataResponse(data=service.to_read(notification))

File: frontend/actions/notifications.ts
"use server";

import type { AppNotification } from "@/types/domain";

import { runAction, runListAction, withServerContext } from "@/lib/server/action-context";
import { requireRole } from "@/lib/server/authorize";
import { toNotification } from "@/lib/server/mappers";
import { parsePagination } from "@/lib/server/pagination";
import { NotificationService } from "@/lib/server/services/notifications";

export type NotificationCreateInput = {
  severity?: string;
  kind?: string;
  title: string;
  message: string;
  recipient?: string | null;
  recipient_email?: string | null;
  entity_type?: string | null;
  entity_id?: string | null;
  metadata?: Record<string, unknown>;
};

export async function listNotifications(params?: {
  page?: number;
  limit?: number;
  unread_only?: boolean;
  kind?: string;
}) {
  return runListAction<AppNotification>(async () => {
    const { client, admin } = await withServerContext();
    const { page, limit } = parsePagination(params?.page, params?.limit);
    const service = new NotificationService(client, admin);
    const { items, total } = await service.list({
      page,
      limit,
      unread_only: params?.unread_only ?? false,
      kind: params?.kind ?? null,
    });
    return { items: items.map(toNotification), total, page, limit };
  });
}

export async function getNotification(id: string) {
  return runAction<AppNotification>(async () => {
    const { client, admin } = await withServerContext();
    const service = new NotificationService(client, admin);
    return toNotification(await service.get(id));
  });
}

export async function createNotification(payload: NotificationCreateInput) {
  return runAction<AppNotification>(async () => {
    const { client, admin, actor } = await withServerContext();
    // Privileged insert bypassing RLS (service-role). Only admins may create
    // notifications from the client; assignment alerts run server-side.
    requireRole(actor, "admin", "create", "notification");
    const service = new NotificationService(client, admin);
    return toNotification(await service.create(payload));
  });
}

export async function markNotificationRead(id: string, read = true) {
  return runAction<AppNotification>(async () => {
    const { client, admin } = await withServerContext();
    const service = new NotificationService(client, admin);
    return toNotification(await service.markRead(id, read));
  });
}

export async function markAllNotificationsRead() {
  return runAction<{ count: number }>(async () => {
    const { client, admin } = await withServerContext();
    const service = new NotificationService(client, admin);
    const count = await service.markAllRead();
    return { count };
  });
}

export async function getUnreadNotificationCount() {
  return runAction<{ count: number }>(async () => {
    const { client, admin } = await withServerContext();
    const service = new NotificationService(client, admin);
    const count = await service.unreadCount();
    return { count };
  });
}